cli/launchpad/model.go

```diff
@@ -0,0 +1,248 @@
+
+// cloudFoundationYAML represents a CloudFoundation CRD.
+//
+// A CloudFoundation CRD can represent a birds-eye view of the entire organization's GCP environment.
+// The CRD can be further broken down into small CRDs (ex: Project, Folder, Org) to represent the same.
+type cloudFoundationYAML struct {
+	commonConfigYAML
+	Spec cloudFoundationSpecYAML `yaml:"spec"`
+}
+
+// cloudFoundationSpecYAML defines CloudFoundation Kind's spec.
+//
+// It is assumed that one CloudFoundation will host at most one Organization.
+type cloudFoundationSpecYAML struct {
+	Org orgSpecYAML `yaml:"organization"`
+}
+
+// UnmarshalYAML evaluates cloudFoundationSpecYAML.
+//
+// UnmarshalYAML will have side effect to push CloudFoundation onto its stack while nested
+// evaluation is ongoing.
+func (c *cloudFoundationSpecYAML) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	gState.push(KindCloudFoundation, c)
+	defer gState.popSilent()
+
+	type raw cloudFoundationSpecYAML
+	if err := unmarshal((*raw)(c)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ==== Organization ====
+
+// orgYAML represents Organization CRD.
+//
+// orgYAML defines a single GCP Organization
+type orgYAML struct {
+	commonConfigYAML
+	Spec orgSpecYAML `yaml:"spec"`
+}
+
+// orgSpecYAML defines Organization Kind's spec.
+type orgSpecYAML struct {
+	Id          string            `yaml:"id"`
+	DisplayName string            `yaml:"displayName"`
+	Folders     []*folderSpecYAML `yaml:"folders"`
+}
+
+// UnmarshalYAML evaluates orgSpecYAML.
+//
+// UnmarshalYAML will have side effect to push Organization onto its stack while nested
+// evaluation is ongoing.
+func (o *orgSpecYAML) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	gState.push(KindOrganization, o)
+	defer gState.popSilent()
+
+	type raw orgSpecYAML
+	if err := unmarshal((*raw)(o)); err != nil {
+		return err
+	}
+	if err := gState.setOrg(o.Id); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ==== Folder ====
+
+// folderYAML represents Folder CRD.
+type folderYAML struct {
+	commonConfigYAML
+	Spec folderSpecYAML `yaml:"spec"`
+}
+
+// folderSpecYAML defines Folder Kind's spec.
+type folderSpecYAML struct { // Inner mappings
+	Id          string                 `yaml:"id"`
+	DisplayName string                 `yaml:"displayName"`
+	ParentRef   parentRefYAML          `yaml:"parentRef"`
+	Folders     []*folderSpecYAML      `yaml:"folders"`
+	Undefined   map[string]interface{} `yaml:",inline"` // Catch-all for untended behavior1
+}
+
+// UnmarshalYAML evaluates folderSpecYAML.
+//
+// UnmarshalYAML will have side effect to push Folder onto its stack while nested
+// evaluation is ongoing. In addition, storing validated folder onto gState for tracking.
+//
+// GCP Folder names required to be in between 3 to 30 characters
+func (f *folderSpecYAML) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	top := gState.peek()
+	gState.push(KindFolder, f)
+	defer gState.popSilent()
+
+	type raw folderSpecYAML
```

Comment:
I keep seeing this pattern in this code. I'm very sure it's wrong. See above for why.